### PR TITLE
Lukaszgryglicki 2346 needs rebase label

### LINKED ISSUE #2346: needs-rebase label not being applied
@fejta noted on kubernetes-dev (https://groups.google.com/d/msg/kubernetes-dev/YNHR4YZa9y4/-bLFgaSRCQAJ) that there was an issue with the needs-rebase label

I noticed that there were PRs with the opposite problem now (actually needs a rebase, but lacks the label):
- https://github.com/kubernetes/kubernetes/pull/42288
- https://github.com/kubernetes/kubernetes/pull/43551

### PR DESCRIPTION
**Pull request by lukaszgryglicki**
Fixes #2346 

But I'm not really happy with this PR and need some feedback.
There are two `FIXME:`'s on this PR also we may need to think about general approach.

Idea was: https://github.com/kubernetes/test-infra/issues/2346#issuecomment-293656429

So I've implemented webhook reacting on PR merging.
Once PR is `(action=closed, merged=true, base.ref=master)` which means merging some PR into master branch - we are using `FindIssues` to find all open, unmerged PRs after waiting for 2 minutes (to give github time to populate `mergeable` flag).
Once we have those PRs we check if their "needs-rebase" label is in sync with `mergeable` flag - eventually add/update label.
I've also added example json payload that can be tested with `cmd/hook` and `phony`

Problems:
- `prow/github/types.go` was missing some `PullRequest` fields. Literally: "Merged" flag (bool), "State" (string), and null-able "Mergeable" - boolptr (nil, *true, *false). `Issue` type had `PullRequest` as `*struct{}` - so it wa sunusable, I've changed it to `*PullRequest` so I had to update many tests to handle that (all of them used PullRequest value as nil or pointer to empty struct, so it was just a cosmetic change)
- `prow/trigger/pr.go` now handles PR merge event, and after that event it waits 2 minutes for GH to populate `mergeable` status and then continues work. I was said that this is ok because plugins run in separate goroutines.
- I was not able to test that against real github api call. I was able to test as `bazel test //prove/...` and via `cmd/hook` as described in `province/README.md` - both of them are not using real GH API call. I've setup HMAC and OAuth2 keys (to run `cmd/hook` without `--local`) but then it requested a lot more config to do locally to run.... I have not cofigured this... so this is my BIGGEST concern here
- Special PR state `pr.State != "skip_sleep"` set in test environment (which cannot happen in real world calls) is used to skip 2 minutes wait (that allows tests to run without delay, and they set mergeable state manually in test cases)
- Label `needs-rebase` is set on the Issue not PR (PR is part of the issue, Labels are defined on the issue), not sure if this is OK.... need confirmation. If not - then our data structure in `prow/github/types.go` is inappropriate: see `Issue` and `PullRequest` structs.
- Last concern is about GH API points..... I think single call to `FindIssues` on each PR merged to master should be just fine (we can call search api up to 30 times/minute), but correct me if I'm wrong.

Summing up.
I'm really waiting for feedback on this one.
We don't had such kind of processing in `prow` bot yet (such kind of calls happen in `mungegithub`) so this is all rthers new, and I';m also very new to `kubernetes` so actually it can happen that all this PR will be discarded.

Sorry for such a long story here, but I'm trying my best to create something useful.

### CODE REVIEW

**Comment by krzyzacy** [comment]
@k8s-bot ok to test
/assign @spxtr @grodrigues3 

please rebase

**Comment by lukaszgryglicki** [comment]
Rebased and fixed conflict

**Comment by cblecker** [comment]
@k8s-bot bazel test this

**Comment by cblecker** [comment]
@lukaszgryglicki Mind taking a look at this further? It looks like it's failing tests on the label plugin.

**Comment by spxtr** [comment]
Ping @lukaszgryglicki

**Comment by lukaszgryglicki** [comment]
Sure, go ahead, I'm working on other stuff atm

**Comment by cblecker** [comment]
Fixed by #3541
/close
